src/logger.coffee

```diff
@@ -5,114 +5,237 @@ es = require 'event-stream'
+		@backend = null
+
+	init: (opts) =>
+		Promise.try =>
+			@opts = opts
+			@_startBackend()
+
+	_startBackend: =>
+		if checkTruthy(@opts.nativeLogger)
+			@backend = new NativeLoggerBackend()
 		else
-			@publish(msg)
+			@backend = new PubnubLoggerBackend()
+		@backend.start(@opts)
+
+	switchBackend: (nativeLogger) =>
+		if checkTruthy(nativeLogger) != checkTruthy(@opts.nativeLogger)
+			@opts.nativeLogger = nativeLogger
+			@backend.stop()
+			@_startBackend()
+
+	enable: (val) =>
+		@opts.enable = val
+		@backend.publishEnabled = checkTruthy(val) ? true
+
+	logDependent: (msg, device) =>
+		@backend.logDependent(msg, device)
+
+	log: (msg) =>
+		@backend.log(msg)
 
-	logSystemMessage: (message, obj, eventName) =>
-		messageObj = { m: message, s: 1 }
-		if obj?.serviceId?
-			messageObj.c = obj.serviceId
+	logSystemMessage: (msg, obj, eventName) =>
+		messageObj = { message: msg, isSystem: true }
+		if obj?.error?
+			messageObj.isStderr = true
 		@log(messageObj)
-		@eventTracker.track(eventName ? message, obj)
+		@eventTracker.track(eventName ? msg, obj)
 
 	lock: (containerId) =>
 		@_writeLock(containerId)
 		.disposer (release) ->
 			release()
 
-	attach: (docker, containerId, serviceId) =>
-		Promise.using @lock(containerId), =>
-			if !@attached[containerId]
+	_attachStream: (docker, stdoutOrStderr, containerId, { serviceId, imageId }) =>
+		Promise.try =>
+			if stdoutOrStderr not in [ 'stdout', 'stderr' ]
+				throw new Error("Invalid log selection #{stdoutOrStderr}")
+			if !@attached[stdoutOrStderr][containerId]
+				logsOpts = { follow: true, stdout: stdoutOrStderr == 'stdout', stderr: stdoutOrStderr == 'stderr', timestamps: true }
 				docker.getContainer(containerId)
-				.logs({ follow: true, stdout: true, stderr: true, timestamps: true })
+				.logs(logsOpts)
 				.then (stream) =>
-					@attached[containerId] = true
+					@attached[stdoutOrStderr][containerId] = true
 					stream
 					.on 'error', (err) =>
 						console.error('Error on container logs', err, err.stack)
-						@attached[containerId] = false
+						@attached[stdoutOrStderr][containerId] = false
 					.pipe(es.split())
 					.on 'data', (logLine) =>
 						space = logLine.indexOf(' ')
 						if space > 0
-							msg = { t: logLine.substr(0, space), m: logLine.substr(space + 1), c: serviceId }
-							@publish(msg)
+							msg = { timestamp: logLine.substr(0, space), message: logLine.substr(space + 1), serviceId, imageId }
+							if stdoutOrStderr == 'stderr'
+								msg.isStderr = true
+							@log(msg)
 					.on 'error', (err) =>
 						console.error('Error on container logs', err, err.stack)
-						@attached[containerId] = false
+						@attached[stdoutOrStderr][containerId] = false
 					.on 'end', =>
-						@attached[containerId] = false
+						@attached[stdoutOrStderr][containerId] = false
+
+	attach: (docker, containerId, serviceInfo) =>
+		Promise.using @lock(containerId), =>
+			@_attachStream(docker, 'stdout', containerId, serviceInfo)
+			.then =>
```

Comment:
Subjective NIT: Can de-indent and keep the promise chain flat

Reply:
Done